<?xml version="1.0"?>
<launch>

    <arg name="use_gui" default="false"/>
    <arg name="use_rviz" default="false"/>
    <arg name="namespace" default="prodigimover"/>
    
    
    <!-- ROS communication between Raspberry Pi 4 and ESP32 (ProdigiMover Base) via TCP-->
    <node pkg="rosserial_python" type="serial_node.py" name="rosserial_prodigimover_node" output="screen">
        <param name="port" value="tcp"/>
        <!--<param name="baud" value="115200"/>-->
    </node>
    
    <!-- Simulation model -->
    <param name="robot_description" command="$(find xacro)/xacro '$(find prodigimover_description)/urdf/prodigimover.urdf'"/>
    <node name="joint_state_publisher" pkg="joint_state_publisher_gui" type="joint_state_publisher_gui" if="$(arg use_gui)"/>
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" unless="$(arg use_gui)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    
    <node pkg="prodigimover_bringup" type="prodigimover_odom_node" name="prodigimover_odom_node" output="screen"/>
    
    <!-- Filter and fuse imu data -->
    <node pkg="imu_filter_madgwick" type="imu_filter_node" name="imu_filter_madgwick" output="screen">
        <param name="fixed_frame" value="base_link"/>
        <param name="use_mag" value="true"/>
        <param name="publish_tf" value="false"/>
        <param name="use_magnetic_field_msg" value="true"/>
        <param name="world_frame" value="enu"/>
        <param name="orientation_stddev" value="0.05"/>
        <param name="angular_scale" value="1.0"/>
        <remap from="/sub_imu" to="/imu/data_raw"/>
        <remap from="/sub_mag" to="/imu/mag"/>
        <remap from="/pub_imu" to="/imu/imu_data"/>
        <remap from="/pub_mag" to="/mag/mag_field"/>
    </node>

    <!-- Post static coordinate system conversion -->
    <node pkg="tf" type="static_transform_publisher" name="base_link_to_imu_link" args="-0.1 0.01 0.01 0 0 0  /base_link /imu_link 100" />

    <!-- Extended Kalman data fusion -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" output="screen">
        <remap from="odometry/filtered" to="odom"/>
        <remap from="/imu0" to="/imu/imu_data"/>
        <remap from="/odom0" to="/raw_odom"/>
        <param name="odom_frame" value="/odom"/>
        <param name="world_frame" value="/odom"/>
        <param name="base_link_frame" value="/robot_footprint"/>
        <rosparam command="load" file="$(find prodigimover_bringup)/param/ekf_robot_localization_new.yaml"/>
    </node>

          
       
    <!-- Handle control node -->
    <include file="$(find prodigimover_teleop)/launch/prodigimover_joy.launch"/>
    
    <!-- rviz 
    <node pkg="rviz" type="rviz" name="odom_rviz" required="true" args="-d $(find prodigimover_bringup)/rviz/bringup.rviz" />-->



    

</launch>
